urllib3/response.py

```diff
@@ -269,7 +270,18 @@ def read(self, amt=None, decode_content=None, cache_content=False):
             except HTTPException as e:
                 # This includes IncompleteRead.
                 raise ProtocolError('Connection broken: %r' % e, e)
+        except Exception:
+            # The response may not be closed but we're not going to use it anymore
+            # so close it now to ensure that the connection is released back to the pool.
+            if self._original_response and not self._original_response.isclosed():
+                self._original_response.close()
 
+            raise
+        finally:
+            if self._original_response and self._original_response.isclosed():
+                self.release_conn()
+
+        if data:
```

Comment:
If no exception is raised during the fp.read opertion, line 279 will never be executed. If we do get an exception it will be caught by one of the inner exception handlers (SocketTimeout, ReadSSLError, HTTPException). All of these re-raise and that will be caught at line 273. It's here that we close the response (but also, this block catches any uncaught exceptions raised in the first block) and then re-raise. The finally block is then executed and we're gone.
If not exception is raised and everything goes according to plan data is set at either line 240 or 244, the finally block is executed and we continue on to decode the data. At least that's what it looks like to me and I'm counting on the test-suite to pick up any blindingly obvious errors with the logic.